apps/etterlatte-behandling/src/main/kotlin/grunnlagsendring/doedshendelse/DoedshendelseService.kt

```diff
@@ -0,0 +1,83 @@
+import no.nav.etterlatte.libs.common.person.Person
+import no.nav.etterlatte.libs.common.person.PersonRolle
+import org.slf4j.LoggerFactory
+import java.time.LocalDate
+import java.time.temporal.ChronoUnit
+import no.nav.etterlatte.libs.common.pdlhendelse.Doedshendelse as PdlDoedshendelse
+
+enum class DoedshendelseFeatureToggle(private val key: String) : FeatureToggle {
+    KanLagreDoedshendelse("pensjon-etterlatte.kan-lage-doedhendelse"),
+    ;
+
+    override fun key(): String = key
+}
+
+class DoedshendelseService(
+    private val doedshendelseDao: DoedshendelseDao,
+    private val pdlTjenesterKlient: PdlTjenesterKlient,
+    private val featureToggleService: FeatureToggleService,
+) {
+    private val logger = LoggerFactory.getLogger(this::class.java)
+
+    fun opprettDoedshendelseForBeroertePersoner(doedshendelse: PdlDoedshendelse) {
+        logger.info("Mottok dødsmelding fra PDL, finner berørte personer og lagrer ned dødsmelding.")
+
+        val avdoed = pdlTjenesterKlient.hentPdlModell(doedshendelse.fnr, PersonRolle.AVDOED, SakType.BARNEPENSJON)
+
+        if (avdoed.doedsdato == null) {
+            sikkerLogg.info("Mottok dødshendelse for ${avdoed.foedselsnummer}, men personen er i live i følge PDL.")
+            logger.info("Mottok dødshendelse fra PDL for en levende person. Avbryter. Se sikkerlogg for detaljer.")
+            return
+        }
+
+        val beroerte = finnBeroerteBarn(avdoed)
+        sikkerLogg.info("Fant ${beroerte.size} berørte personer for avdød (${avdoed.foedselsnummer})")
+
+        lagreDoedshendelse(beroerte, avdoed)
+    }
+
+    private fun lagreDoedshendelse(
+        beroerte: List<Person>,
+        avdoed: PersonDTO,
+    ) {
+        if (featureToggleService.isEnabled(DoedshendelseFeatureToggle.KanLagreDoedshendelse, false)) {
+            inTransaction {
+                beroerte.forEach { barn ->
+                    doedshendelseDao.opprettDoedshendelse(
+                        Doedshendelse.nyHendelse(
+                            avdoedFnr = avdoed.foedselsnummer.verdi.value,
+                            avdoedDoedsdato = avdoed.doedsdato!!.verdi,
+                            beroertFnr = barn.foedselsnummer.value,
+                            relasjon = Relasjon.BARN,
+                        ),
+                    )
+                }
+            }
+        }
+    }
+
+    private fun finnBeroerteBarn(avdoed: PersonDTO): List<Person> {
+        val maanedenEtterDoedsfall = avdoed.doedsdato!!.verdi.plusMonths(1).withDayOfMonth(1)
+
+        return with(avdoed.avdoedesBarn ?: emptyList()) {
+            this.filter { barn -> barn.doedsdato == null }
+                .filter { barn -> barn.under20PaaDato(maanedenEtterDoedsfall) }
+        }
+    }
+}
+
+private fun Person.under20PaaDato(dato: LocalDate): Boolean =
+    if (foedselsdato != null) {
+        ChronoUnit.YEARS.between(foedselsdato, dato) < 20
+    } else {
+        foedselsnummer.getAgeAtDate(dato) < 20
```

Comment:
Jeg vet ikke om denne gir mening -- hvis vi ikke har fødselsdato vi kan stole på er det nok noe off med identifikator-oppsett også. Se f.eks. håndtering hvis man mangler tilgjengelige d-nummer for en fødselsdato: https://www.udi.no/ord-og-begreper/d-nummer/#link-32990

Reply:
Hmm, ser nå at den `getAge`-metoden vi har på `Folkeregistertidentifikator` kun har blitt benyttet ifm. migrering. Det tryggeste er kanskje å regne de som mangler fødselsdato som "potensielle" berørte personer som må sjekkes opp manuelt da.

Reply:
https://pdl-docs.intern.nav.no/ekstern/index.html#_f%C3%B8dsel har jo litt ekstra info -- mulig at vi kan benytte fødselsår for å redusere antall tvilstilfeller i alle fall, siden det er garantert

Reply:
[b896a3f](https://github.com/navikt/pensjon-etterlatte-saksbehandling/pull/3772/commits/b896a3fde403881e5c1f8fa70e500cba564b2d1e)